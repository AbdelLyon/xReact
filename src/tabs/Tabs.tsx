import { Children, isValidElement, PropsWithChildren } from "react";

import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/ui/card";
import {
  Tabs as TabsRoot,
  TabsContent,
  TabsList,
  TabsTrigger,
} from "@/ui/tabs";
import { cn } from "@/utils";

export interface TabDefinition {
  value: string;
  titleTrigger: string;
  title?: string;
  description?: string;
}

interface ClassNames {
  root?: string;
  tabsList?: string;
  tabsTrigger?: string;
  card?: string;
  cardHeader?: string;
  cardTitle?: string;
  cardDescription?: string;
  cardContent?: string;
}

type TabsProps = PropsWithChildren<{
  defaultTab: string;
  tabs: TabDefinition[];
  classNames?: ClassNames;
  variant?: "default" | "button";
  onValueChange?: (value: string) => void;
}>;

function Tabs({
  defaultTab,
  tabs,
  classNames = {},
  children,
  variant,
  onValueChange,
}: Readonly<TabsProps>) {
  return (
    <TabsRoot
      className={classNames.root}
      defaultValue={defaultTab}
      onValueChange={onValueChange}
    >
      <TabsList className={cn("flex w-max gap-2", classNames.tabsList)}>
        {tabs.map((tab) => (
          <TabsTrigger
            key={tab.value}
            className={classNames.tabsTrigger}
            value={tab.value}
            variant={variant}
          >
            {tab.titleTrigger}
          </TabsTrigger>
        ))}
      </TabsList>
      {Children.map(children, (child) => {
        if (isValidElement(child) && child.type === TabsContent) {
          const tab = tabs.find((t) => t.value === child.props.value);

          return (
            <TabsContent key={child.props.value} value={child.props.value}>
              <Card className={classNames.card}>
                {(tab?.title || tab?.description) && (
                  <CardHeader className={classNames.cardHeader}>
                    {tab?.title && (
                      <CardTitle className={classNames.cardTitle}>
                        {tab.title}
                      </CardTitle>
                    )}
                    {tab?.description && (
                      <CardDescription className={classNames.cardDescription}>
                        {tab.description}
                      </CardDescription>
                    )}
                  </CardHeader>
                )}

                <CardContent
                  className={cn("space-y-2", classNames.cardContent)}
                >
                  {child}
                </CardContent>
              </Card>
            </TabsContent>
          );
        }

        return null;
      })}
    </TabsRoot>
  );
}

export { Tabs, TabsContent };
