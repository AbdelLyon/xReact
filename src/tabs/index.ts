import "@/index.css";
export { Tabs, TabsContent, type TabDefinition } from "@/tabs/Tabs";
