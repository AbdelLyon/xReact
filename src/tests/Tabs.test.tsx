import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { Tabs, TabsContent } from "@/tabs";

// Mock des dÃ©pendances
type Props = { children: React.ReactNode; [key: string]: any };
vi.mock("@/ui/card", () => ({
  Card: ({ children, ...props }: Props) => (
    <div data-testid="card" {...props}>
      {children}
    </div>
  ),
  CardContent: ({ children, ...props }: Props) => (
    <div data-testid="card-content" {...props}>
      {children}
    </div>
  ),
  CardDescription: ({ children, ...props }: Props) => (
    <div data-testid="card-description" {...props}>
      {children}
    </div>
  ),
  CardHeader: ({ children, ...props }: Props) => (
    <div data-testid="card-header" {...props}>
      {children}
    </div>
  ),
  CardTitle: ({ children, ...props }: Props) => (
    <div data-testid="card-title" {...props}>
      {children}
    </div>
  ),
}));

vi.mock("@/ui/tabs", () => ({
  Tabs: ({ children, ...props }: Props) => (
    <div data-testid="tabs" {...props}>
      {children}
    </div>
  ),
  TabsContent: ({ children, ...props }: Props) => (
    <div data-testid="tabs-content" {...props}>
      {children}
    </div>
  ),
  TabsList: ({ children, ...props }: Props) => (
    <div data-testid="tabs-list" {...props}>
      {children}
    </div>
  ),
  TabsTrigger: ({ children, ...props }: Props) => (
    <div data-testid="tabs-trigger" {...props}>
      {children}
    </div>
  ),
}));

describe("Tabs Component", () => {
  it("renders correctly", () => {
    const tabsConfig = [
      {
        value: "tab1",
        title: "Tab 1",
        description: "Description for Tab 1",
        titleTrigger: "Tab 1",
      },
      {
        value: "tab2",
        title: "Tab 2",
        description: "Description for Tab 2",
        titleTrigger: "Tab 2",
      },
    ];

    const { asFragment } = render(
      <Tabs defaultTab="tab1" tabs={tabsConfig}>
        <TabsContent value="tab1">
          <p>Content for Tab 1</p>
        </TabsContent>
        <TabsContent value="tab2">
          <p>Content for Tab 2</p>
        </TabsContent>
      </Tabs>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it("renders with custom classNames", () => {
    const tabsConfig = [
      {
        value: "tab1",
        title: "Tab 1",
        description: "Description for Tab 1",
        titleTrigger: "Tab 1 Trigger",
      },
    ];

    const customClassNames = {
      root: "custom-root",
      tabsList: "custom-tabs-list",
      tabsTrigger: "custom-tabs-trigger",
      card: "custom-card",
      cardHeader: "custom-card-header",
      cardTitle: "custom-card-title",
      cardDescription: "custom-card-description",
      cardContent: "custom-card-content",
    };

    const { asFragment } = render(
      <Tabs classNames={customClassNames} defaultTab="tab1" tabs={tabsConfig}>
        <TabsContent value="tab1">
          <p>Content for Tab 1</p>
        </TabsContent>
      </Tabs>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
