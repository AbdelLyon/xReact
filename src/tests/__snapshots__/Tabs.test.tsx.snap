// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs Component > renders correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="tabs"
  >
    <div
      class="flex w-max gap-2"
      data-testid="tabs-list"
    >
      <div
        data-testid="tabs-trigger"
        value="tab1"
      >
        Tab 1
      </div>
      <div
        data-testid="tabs-trigger"
        value="tab2"
      >
        Tab 2
      </div>
    </div>
    <div
      data-testid="tabs-content"
      value="tab1"
    >
      <div
        data-testid="card"
      >
        <div
          data-testid="card-header"
        >
          <div
            data-testid="card-title"
          >
            Tab 1
          </div>
          <div
            data-testid="card-description"
          >
            Description for Tab 1
          </div>
        </div>
        <div
          class="space-y-2"
          data-testid="card-content"
        >
          <div
            data-testid="tabs-content"
            value="tab1"
          >
            <p>
              Content for Tab 1
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="tabs-content"
      value="tab2"
    >
      <div
        data-testid="card"
      >
        <div
          data-testid="card-header"
        >
          <div
            data-testid="card-title"
          >
            Tab 2
          </div>
          <div
            data-testid="card-description"
          >
            Description for Tab 2
          </div>
        </div>
        <div
          class="space-y-2"
          data-testid="card-content"
        >
          <div
            data-testid="tabs-content"
            value="tab2"
          >
            <p>
              Content for Tab 2
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs Component > renders with custom classNames 1`] = `
<DocumentFragment>
  <div
    class="custom-root"
    data-testid="tabs"
  >
    <div
      class="flex w-max gap-2 custom-tabs-list"
      data-testid="tabs-list"
    >
      <div
        class="custom-tabs-trigger"
        data-testid="tabs-trigger"
        value="tab1"
      >
        Tab 1 Trigger
      </div>
    </div>
    <div
      data-testid="tabs-content"
      value="tab1"
    >
      <div
        class="custom-card"
        data-testid="card"
      >
        <div
          class="custom-card-header"
          data-testid="card-header"
        >
          <div
            class="custom-card-title"
            data-testid="card-title"
          >
            Tab 1
          </div>
          <div
            class="custom-card-description"
            data-testid="card-description"
          >
            Description for Tab 1
          </div>
        </div>
        <div
          class="space-y-2 custom-card-content"
          data-testid="card-content"
        >
          <div
            data-testid="tabs-content"
            value="tab1"
          >
            <p>
              Content for Tab 1
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
